hypothesis-python/src/hypothesis/extra/ghostwriter.py

```diff
@@ -0,0 +1,309 @@
+    # We use `infer` and go via `builds()` instead of directly through
+    # `from_type()` so that `get_type_hints()` can resolve any forward
+    # references for us.
+    if param.annotation is not inspect.Parameter.empty:
+        return infer
+    # If there's no annotation and no default value, the user will have to
+    # fill this in later.  We use nothing() as a placeholder to this effect.
+    if param.default is inspect.Parameter.empty:
+        return st.nothing()
+    # If our default value is an Enum or a boolean, we assume that any value
+    # of that type is acceptable.  Otherwise, we only generate the default.
+    if isinstance(param.default, bool):
+        return st.booleans()
+    if isinstance(param.default, enum.Enum):
+        return st.sampled_from(type(param.default))
+    return st.just(param.default)
+
+
+def _get_params(func: Callable) -> Dict[str, inspect.Parameter]:
+    """Get non-vararg parameters of `func` as an ordered dict."""
+    var_param_kinds = (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+    params = list(inspect.signature(func).parameters.values())
+    return OrderedDict((p.name, p) for p in params if p.kind not in var_param_kinds)
+
+
+def _get_strategies(
+    *funcs: Callable, pass_result_to_next_func: bool = False
+) -> Dict[str, st.SearchStrategy]:
+    """Return a dict of strategies for the union of arguments to `funcs`.
+
+    If `pass_result_to_next_func` is True, assume that the result of each function
+    is passed to the next, and therefore skip the first argument of all but the
+    first function.
+
+    This dict is used to construct our call to the `@given(...)` decorator.
+    """
+    given_strategies = {}  # type: Dict[str, st.SearchStrategy]
+    for i, f in enumerate(funcs):
+        params = _get_params(f)
+        if pass_result_to_next_func and i >= 1:
+            del params[next(iter(params))]
+        builder_args = {k: _strategy_for(v) for k, v in params.items()}
+        strat = st.builds(f, **builder_args).wrapped_strategy  # type: ignore
+
+        args, kwargs = strat.mapped_strategy.wrapped_strategy.element_strategies
+        if args.element_strategies:
+            raise NotImplementedError("Expected to pass everything as kwargs")
+
+        for k, v in zip(kwargs.keys, kwargs.mapped_strategy.element_strategies):
+            if isinstance(v, OneOfStrategy):
+                # repr nested one_of as flattened (their real behaviour)
+                v = st.one_of(v.element_strategies)
+            if repr(given_strategies.get(k, v)) != repr(v):
+                given_strategies[k] = given_strategies[k] | v
+            else:
+                given_strategies[k] = v
+
+    # If there is only one function, we pass arguments to @given in the order of
+    # that function's signature.  Otherwise, we use alphabetical order.
+    if len(funcs) == 1:
+        indexer = list(_get_params(f)).index
```

Comment:
Up to you whether you change this but I'd almost always write something like `{v: i for i in enumerate(_get_params(f))}`. Slightly longer but a lot more obvious and (this doesn't matter in this case) doesn't involve an `O(n)` lookup.

Reply:
Or, now that I'm actually looking at it again, `{name: given_strategies[name] for name in _get_params(f)}` :facepalm: